### ISSUE TITLE
Question about `pkgs` attribute

### ISSUE THREAD

**Issue by musjj** [Misterio77/nix-starter-configs#38]
I'm slightly confused about the handling of `nixpkgs` in flake configs. 

The flake introduced an input for `nixpkgs`:

https://github.com/Misterio77/nix-starter-configs/blob/9e4963ae5c1130b18a64db2790248e75a11f94cb/minimal/flake.nix#L4-L6

And we override the system's registery with:

https://github.com/Misterio77/nix-starter-configs/blob/9e4963ae5c1130b18a64db2790248e75a11f94cb/minimal/nixos/configuration.nix#L38-L45

Does that mean that the `pkgs` attribute that a module receives is now using the new `nixpkgs` introduced in our flake? Or am I misunderstanding something here.

**Comment by Misterio77** [comment]
Hey,

The **registry** and module **arguments** are different things.

`nix.registry` sets the system flake registry, that is, what flake references (e.g. `nixpkgs`) resolve to, usually on the CLI (there's also a less common case, that is when you e.g. use `nixpkgs` as argument on the flake outputs function without defining it on the inputs first).

The arguments (by default `nixpkgs`, `lib`, `config`) a NixOS module receives are not affected by the flake registry at all.  This template passes `inputs` and `outputs` (via `specialArgs`/`extraSpecialArgs`), it does not pass `nixpkgs` at all (it can be accessed through `inputs.nixpkgs` though).

This can sound a little confusing, but `nixpkgs` usually refers to the repo and `pkgs` refers to an _instanced_ version of `nixpkgs` (e.g. one that is configured with a `system`, optionally with `overlays` and that sort of thing). The `pkgs` a nixos module receives is created by taking the `nixpkgs` repo where `nixosConfiguration` came from, and instancing it with `nixpkgs.*` options you have set (e.g. `nixpkgs.overlays`, `nixpkgs.hostPlatform`). You can change which `nixpkgs` is used to instance `pkgs` by using your preferred `nixpkgs`'s `nixosSytem`; for example, you could add a `nixpkgs-stable` repo to your inputs and use `nixpkgs-stable.nixosConfiguration`.

Long story short: no. The only ways the flakes can your configuration is on the `nixosConfiguration` call itself; in the case of this template, all it does is pass `inputs` and `outputs` as extra arguments, `pkgs` is still created by the nixos module system.

Does that make sense? Hope I could make things a little easier to understand. Let me know if something's not clear, and I can try and explain it another way!

**Comment by musjj** [comment]
Thank you, this is pretty enlightening! So the "environment" of your system is separate from the one used when you're building it, is that right (not sure of the terms)? 

So what this template config does is that it registers the flakes (from inputs) used when building the system into the system "environment" itself, so that it can be used from the CLI?

Another question: what's the difference between a system and global flake? When I run `nix register list`, this is what I get:

```
system flake:nixpkgs path:/nix/store/{some-really-long-hash}
system flake:self path:/nix/store/{some-really-long-hash}
global flake:agda github:agda/agda
...
global flake:nixpkgs github:NixOS/nixpkgs/nixpkgs-unstable
...
```

It seems that there are two `nixpkgs` flake, a system and a global one. What's the difference? Also what's the `self` flake?

Another thing I'm noticing is that it seems that the `nix.registry` and `nix.nixPath` options are *imperative*. So even if I remove my registry options and rebuild, the registry remains the same.

**Comment by Misterio77** [comment]
> Thank you, this is pretty enlightening!

No prob!

> So the "environment" of your system is separate from the one used when you're building it, is that right (not sure of the terms)?

In a way, yeah. There's stuff being used to compute other stuff, and not necessarily all of the first stuff is used by you directly after the build.

> So what this template config does is that it registers the flakes (from inputs) used when building the system into the system "environment" itself, so that it can be used from the CLI?

Yup. Your flake inputs get copied to the nix store, and your `system` (more on these below) registries are set to be the path of the inputs, so you get the same flakes on both the CLI and on the system itself.

> Also what's the self flake?

One of the `inputs` is actually `self`. Look at this:
```nix
{
  inputs = {
    # ...
  };
  outputs = { self, nixpkgs } @ inputs: {
    # ...
  };
}
```

You might use `{ nixpkgs, ... } @ inputs` or similar, but `inputs.self` is still there. So, by passing `inputs` to `nix.registry`, `self` becomes a registry entry like all other inputs. `self` is the same flake (i.e. your configuration repo). It's pretty useful, you can do stuff like `nix shell self#nixosConfigurations.foo.pkgs.hello` (this works even if you have overlays, custom packages, etc, so pretty useful) or `nix eval self#nixosConfigurations.foo.config.networking.hostName`.

> Another question: what's the difference between a system and global flake?

[There's three different registries: the global registry, the system registry, and the user registry.](https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-registry.html) All of those work the same way (they map an alias to any sort of flake you want, be it a repo on specific commit, a repo on the main branch, a path, etc). They are, however, configured in different ways; in order of increasing priority:

- The global registry. The lower precedence registry is there to make the out-of-box experience of using nix a little better. Your `/etc/nix.conf` has an option called `flake-registry`, it is a URL pointing to a JSON that nix will download and use as registry - by default, it's this one: https://github.com/NixOS/flake-registry/blob/master/flake-registry.json. If I'm setting the system registry, I like to disable this one, which you can do by setting `flake-registry` to an empty URL (on NixOS configs you can simply use `nix.settings.flake-registry = ""`).
- The system registry. It is located at `/etc/nix/registry.json`; we can configure it straight by editing the file or through NixOS configs using `nix.registry` (which is what this flake does).
- The user registry. It is located at `~/.config/nix/registry.json`. This one can be imperatively manipulated, yes (it's mutated when you use `nix registry add`/`pin`/`remove`, but you can also use home-manager to set it similarly to the system one.

So, no, they're not actually entirely imperative, not fully:
- The registry has three layers, one of them is either a file fetched from somewhere or nothing at all, the second one is configured on your system declaratively (even on non-NixOS, it also comes from a file that is not mutated by the CLI), the third one is configured on your users either declaratively (through the file or with home-manager) or imperatively (mutable by the commands).
- Nix path is whole different beast: you have channels (which are the imperative, per-user, part), and you can also set the _environment variable_ `NIX_PATH` (this is kinda of a hack to get it to be declarative, and [is what `nix.nixPath` sets](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/nix-channel.nix#L95)).

You can see that flakes+registries (minus the global one, I think it should be as big as it is, personally) are a cleaner solution to what nix path/channels were originally created for.

I see that there's two main ways someone uses nix: in an ad-hoc way (i.e. you don't care much about the version of a package, you just want it in your shell now), and in a explicit way (i.e. for a project other people will use). Channels don't work well for the first one (as they don't auto-update) neither the second one (they are not declaratively set on the project). Registries are great for the first case (even letting you consume a different commit or branch on the CLI when needed), and flakes are great for the second case.

**Comment by musjj** [comment]
That cleared up all my doubts, your posts really helped me develop a better mental model of Nix and NixOS!
Thanks for your help and for making this wonderful repo!
